core/BranchPred.hpp

```diff
@@ -0,0 +1,96 @@
+// <Branch.hpp> -*- C++ -*-
+
+//!
+//! \file BranchPred.hpp
+//! \brief Definition of Branch Prediction API
+//!
+
+/*
+ * This file defines the Branch Prediction API.
+ * The goal is to define an API that is generic and yet flexible enough to support various
+ * branch prediction microarchitecture.
+ * To the end, we envision a generic branch predictor as a black box with following inputs
+ * and outputs:
+ *   * A generic Prediction output
+ *   * A generic Prediction input
+ *   * A generic Update input
+ *
+ * The generic branch predictor may have two operations:
+ *   * getPrediction: produces Prediction output based on the Prediction input.
+ *   * updatePredictor: updates Predictor with Update input.
+ *
+ * It is intended that an implementation of branch predictor must also specify
+ * implementations of Prediction output, Prediction input and Update input, along with
+ * implementations of getPrediction and updatePredictor operations.
+ * */
+#pragma once
+
+#include <cstdint>
+#include <map>
+#include "sparta/utils/SpartaAssert.hpp"
+
+template <class PredictionT, class UpdateT, class InputT>
+class BranchPredictorIF
+{
+public:
+    virtual PredictionT getPrediction(const InputT &) = 0;
+    virtual void updatePredictor(UpdateT &) = 0;
+};
+
+// following class definitions are example inputs & outputs for a very simple branch
+// predictor
+class DefaultPrediction
+{
+public:
+    // index of branch instruction in the fetch packet
+    // branch_idx can vary from 0 to (FETCH_WIDTH - 1)
+    uint32_t branch_idx;
+    // predicted target PC 
+    uint64_t predictedPC;
```

Comment:
inconsistent coding style. Camelcase is for function names, not memeber variables
use: predicted_PC

Reply:
💯 

Also should initialized to defaults to catch errors:
```
    uint32_t branch_idx = std::numeric_limits<uint32_t>::max();
    // predicted target PC 
    uint64_t predictedPC = std::numeric_limits<uint64_t>::max();
```